lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation.hs

```diff
@@ -771,6 +792,19 @@ class Show from => MkKeyFingerprint (key :: Depth -> Type -> Type) from where
 data ErrMkKeyFingerprint key from
     = ErrInvalidAddress from (Proxy key) deriving (Show, Eq)
 
+data AddressParts = AddressParts
+    { addrType :: Word8
+    , addrNetwork :: Word8
+    , rest :: ByteString
+    } deriving (Show,Eq)
+
+toAddressParts :: Address -> AddressParts
+toAddressParts (Address bytes) = AddressParts {..}
+  where
+    (fstByte, rest) = first BS.head $ BS.splitAt 1 bytes
+    addrType = fstByte .&. 0b11110000
+    addrNetwork = fstByte .&. 0b00001111
+
```

Comment:
(Longer term thought:) I wonder if not using ledger types (in e.g. isOurs) would be cleaner when the required differentiation between payment and script credentials already exists in the types we first get served from the node 🤔

Reply:
this is reused from cardano-addresses. in the first byte there is information for address type (first 4 bits), and network (4 last bits.)

Reply:
I know. But instead of

```haskell
    paymentKeyFingerprint addr =
         let AddressParts{..} = toAddressParts addr
             baseAddr = 0b00110000       -- scripthash; scripthash
             enterpriseAddr = 0b01110000 -- scripthash
             rewardAcct = 0b11110000     -- scripthash
         in if addrType `elem` [baseAddr, enterpriseAddr, rewardAcct] then
             Right $ KeyFingerprint $ BS.take hashSize rest
            else
             Left $ ErrInvalidAddress addr (Proxy @SharedKey)
```

you could have

```haskell
    paymentKeyFingerprint addr = case unsafeDeserializeLedgerAddr addr of
            Ledger.Addr _networkId (ScriptHashObj paymentScriptHash) _ -> Right $ Fingerprint paymentScriptHash
            _ -> Left $ ErrInvalidAddress addr (Proxy @SharedKey)
```

which has the benefit of automatically including Byron addresses in the `_` case.

And more long term, we could drop the `unsafeDeserializeLedgerAddr` if we dropped the conversion _to_ `W.Address`.

Reply:
frankly speaking these are equivalent approaches. One is just looking at first byte and decomposing it on addr type and network tag, the second is relying on ledger code to do this and exposing it through pattern matching. I am slightly in favor of the approach I adopted for two reasons: 1. this is the approach from cardano-addresses and we want very soon to reuse these parts in cardano-wallet. 2. if it was near Compatibility modules where ledger types are omnipresent I would be more eager. why? they use shortByte for hashes and we would need to use fromShort and toShort during that address analysis.